File: backend/meetup/event/src/main/java/org/zil/event/EventSvc.java
package org.zil.event;

import io.github.resilience4j.retry.annotation.Retry;
import lombok.AllArgsConstructor;
import org.springframework.amqp.core.Queue;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.dao.DuplicateKeyException;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.zil.event.dto.user.XValidUserRes;
import org.zil.event.request.CreateEventReq;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

@Service
@AllArgsConstructor
public class EventSvc {

    private final EventRepo eventRepo;
    private final RestTemplate mrest;
    private final RabbitTemplate rabbitTemplate;
    private final EventConfig config;

    @Retry(name = "create-event", fallbackMethod = "createEventFallback")
    public boolean create(CreateEventReq req, String currentUserEmail, String authToken) {

        if (currentUserEmail == null)
            throw new IllegalStateException("Can't handle this request, unable to resolve authorized user for this request");

        if(!req.getStartAt().isAfter(LocalDateTime.now().minusSeconds(1L)))
            throw new IllegalStateException("Event start date cannot be behind the current time");

        XValidUserRes res = null;
        try{
            HttpHeaders headers = new HttpHeaders();
            headers.set(HttpHeaders.AUTHORIZATION, "Bearer "+authToken);
            HttpEntity<Void> requestEntity = new HttpEntity<>(headers);

            res = mrest.exchange(
                    config.GatewayBaseUri+"/api/v1/user/isvalid/email/{userId}",
                    HttpMethod.GET,
                    requestEntity,
                    XValidUserRes.class,
                    currentUserEmail).getBody();
        } catch (Exception e) {
            throw new IllegalStateException("Unable to continue with request, owning user can't be verified");
        }

        if (res == null)
            throw new RuntimeException("Couldn't verify owning user genuinely");

        if (!res.exist())
            throw new IllegalStateException("The owning user doesn't exist in the system");

        if (eventRepo.findByTitleAndOwnership(req.getTitle(), res.id()).isPresent())
            throw new DuplicateKeyException("Event already in use");

        //check if ownership has another event during same duration
        Date start = Date.from(req.getStartAt().atZone(ZoneId.of("Africa/Lagos")).toInstant());
        if (eventRepo.findByOwnershipAndDateInBetweenDates(res.id(), start).size() > 0)
            throw new IllegalStateException("One of your event is existing within time window specified in start date");

        res = null;



        Date end = Date.from(req.getStartAt().plusMinutes(req.getDuration()).atZone(ZoneId.of("Africa/Lagos")).toInstant());
//        Event ev = Event.builder()
//                .ownership(res.id())
//                .title(req.getTitle())
//                .duration(req.getDuration())
//                .startAt(start)
//                .endAt(end)
//                .createdAt(new Date())
//                .build();
//
//        ev.setModifiedAt();
//
//        eventRepo.saveAndFlush(ev);

        Map<String, String> objectMap = new HashMap<>();
        objectMap.put("message", "Now you see me");
        objectMap.put("type", "EMAIL");
        objectMap.put("target", "akindutire33@gmail.com");

        rabbitTemplate.convertAndSend(config.NOTIF_QUEUE, objectMap);
        return true;
    }

    public Boolean createEventFallback (Exception e) {
        return false;
    }
}

File: backend01/user.event/src/main/kotlin/com/meet/user/event/service/domain/EventSvc.kt
package com.meet.user.event.service.domain

import com.meet.user.event.contract.domain.EventCt
import com.meet.user.event.dao.EventRepo
import com.meet.user.event.dto.EventDto
import com.meet.user.event.entity.Event
import com.meet.user.event.request.CreateEventReq
import com.meet.user.event.request.EditEventReq
import feign.FeignException
import org.springframework.beans.BeanUtils
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.*


@Component
class EventSvc:EventCt {

    lateinit var eventRepo: EventRepo

    override fun create(req: CreateEventReq) : EventDto {

        var e = Event()
        e.ev_token = UUID.randomUUID().toString()
        e.period = req.expectedPeriod as Int
        e.type = req.type
        e.capacity = req.expectedCapacity as Int
        e.description = req.description.toString()
        e.createdAt = LocalDateTime.now()
        e.updatedAt = LocalDateTime.now()

        e.timezone = ""

        e = eventRepo.save(e)
        val et = EventDto()

        BeanUtils.copyProperties(e, et)
        return et
    }

    override fun edit(req: EditEventReq) : EventDto {

        val timezone = ""

        val e: Event = eventRepo.findByEvToken(req.evToken)

        e.updatedAt = LocalDateTime.now().atZone(ZoneId.of(timezone)).toLocalDateTime()
        e.period = req.expectedPeriod as Int;
        e.capacity = req.expectedCapacity as Int
        e.description = req.description.toString()

        val et = EventDto()
        BeanUtils.copyProperties(e, et)
        return et
    }
}